include/drivers/rtio.h

```diff
@@ -0,0 +1,486 @@
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s64_t);
+
+struct rtio_block_allocator;
+
+/**
+ * @private
+ * @brief Function definition for allocating rtio blocks from a given allocator
+ */
+typedef int (*rtio_block_alloc_t)(struct rtio_block_allocator *allocator,
+				  struct rtio_block **block, size_t size,
+				  u32_t timeout);
+
+/**
+ * @private
+ * @brief Function definition for freeing rtio blocks
+ */
+typedef void (*rtio_block_free_t)(struct rtio_block_allocator *allocator,
+				  struct rtio_block *block);
+
+/**
+ * @brief An rtio block allocator interface
+ */
+struct rtio_block_allocator {
+	rtio_block_alloc_t alloc;
+	rtio_block_free_t free;
+};
+
+/**
+ * @brief An rtio mempool block allocator
+ */
+struct rtio_mempool_block_allocator {
+	struct rtio_block_allocator allocator;
+	struct k_mem_pool *mempool;
+};
+
+/**
+ * @brief An rtio_block allocated from a mempool
+ */
+struct rtio_mempool_block {
+	struct rtio_block block;
+	struct k_mem_block_id id;
+};
+
+/**
+ * @brief Allocate a block from a mempool
+ */
+static inline int rtio_mempool_block_alloc(
+	struct rtio_block_allocator *allocator, struct rtio_block **block,
+	size_t size, u32_t timeout)
+{
+	struct rtio_mempool_block_allocator *mempool_allocator =
+		(struct rtio_mempool_block_allocator *)allocator;
+	struct k_mem_block memblock;
+	size_t block_size = size + sizeof(struct rtio_mempool_block);
+
+	int res = k_mem_pool_alloc(mempool_allocator->mempool,
+				   &memblock, block_size,
+				   timeout);
+	if (res == 0) {
+		struct rtio_mempool_block *mempool_block =
+			(struct rtio_mempool_block *)(memblock.data);
+		size_t struct_size = sizeof(struct rtio_block);
+		u8_t *dataptr = &((u8_t *)memblock.data)[struct_size];
+
+		mempool_block->id = memblock.id;
+		*block = (struct rtio_block *)mempool_block;
+		rtio_block_init((struct rtio_block *)*block, dataptr, size);
+	}
+	return res;
+}
+
+static inline void rtio_mempool_block_free(
+	struct rtio_block_allocator *allocator,
+	struct rtio_block *block)
+{
+	struct rtio_mempool_block *mempool_block =
+		(struct rtio_mempool_block *)block;
+
+	k_mem_pool_free_id(&mempool_block->id);
+}
+
+/**
+ * @brief Define an rtio block mempool allocator
+ */
+#define RTIO_MEMPOOL_ALLOCATOR_DEFINE(_name, _minsz, _maxsz, _nmax)	\
+	K_MEM_POOL_DEFINE(rtio_pool_mempool_##_name,			\
+			  sizeof(struct rtio_mempool_block) + _minsz,	\
+			  sizeof(struct rtio_mempool_block) + _maxsz,	\
+			  _nmax,					\
+			  4);						\
+									\
+	struct rtio_mempool_block_allocator _name = {			\
+	.allocator = {							\
+	.alloc = rtio_mempool_block_alloc,				\
+	.free = rtio_mempool_block_free					\
+},									\
+	.mempool = &rtio_pool_mempool_##_name				\
+}
+
+
+/**
+ * @brief Allocate a rtio_block from an rtio_pool with a timeout
+ *
+ * This call is not safe to do with a timeout other than K_NO_WAIT
+ * in an interrupt handler.
+ *
+ * The allocator *must* have a memory layout equivalent to
+ * struct rtio_block_allocator
+ */
```

Comment:
Here and all other public API: document the parameters and the return values, including expected error codes.  Particularly `size` which is presumably in bytes, not blocks (which need not be fixed size).  (A new user of the API will not know and could get confused.)